### ISSUE TITLE
在train.py中取消注释tb_writer.add_graph(model, imgs) ，会报错。注释掉则正常运行

### ISSUE THREAD

**Issue by wyuzyf** [ultralytics/yolov5#2284]
## ❔Question
在train.py中取消注释tb_writer.add_graph(model, imgs) ，会报错

EEROR如下：

## Additional context

module._c._create_method_from_trace(method_name, func, example_inputs, var_lookup_fn, strict, _force_outplace)
D:\zhaoyinfa\code\zdws\models\yolo.py:45: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if self.grid[i].shape[2:4] != x[i].shape[2:4]:
     0/299        0G    0.1403    0.0758         0    0.2161        24       640:   0%|          | 0/17 [00:05<?, ?it/s]
Traceback (most recent call last):
  File "D:/zhaoyinfa/code/zdws/train.py", line 442, in <module>
    train(hyp, opt, device, tb_writer)
  File "D:/zhaoyinfa/code/zdws/train.py", line 294, in train
    tb_writer.add_graph(model, imgs)  # add model to tensorboard
  File "D:\ProgramData\Anaconda3\lib\site-packages\torch\utils\tensorboard\writer.py", line 714, in add_graph
    self._get_file_writer().add_graph(graph(model, input_to_model, verbose))
  File "D:\ProgramData\Anaconda3\lib\site-packages\torch\utils\tensorboard\_pytorch_graph.py", line 285, in graph
    trace = torch.jit.trace(model, args)
  File "D:\ProgramData\Anaconda3\lib\site-packages\torch\jit\__init__.py", line 955, in trace
    check_tolerance, strict, _force_outplace, _module_class)
  File "D:\ProgramData\Anaconda3\lib\site-packages\torch\jit\__init__.py", line 1119, in trace_module
    check_tolerance, strict, _force_outplace, True, _module_class)
  File "D:\ProgramData\Anaconda3\lib\site-packages\torch\autograd\grad_mode.py", line 15, in decorate_context
    return func(*args, **kwargs)
  File "D:\ProgramData\Anaconda3\lib\site-packages\torch\jit\__init__.py", line 742, in _check_trace
    raise TracingCheckError(*diag_info)
torch.jit.TracingCheckError: Tracing failed sanity checks!
ERROR: Graphs differed across invocations!
	Graph diff:
		  graph(%self.1 : __torch__.models.yolo.Model,
		        %x : Tensor):
		    %2 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %3 : __torch__.models.yolo.Detect = prim::GetAttr[name="24"](%2)
		    %4 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %5 : __torch__.models.common.BottleneckCSP = prim::GetAttr[name="23"](%4)
		    %6 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %7 : __torch__.models.common.Concat = prim::GetAttr[name="22"](%6)
		    %8 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %9 : __torch__.models.common.Conv = prim::GetAttr[name="21"](%8)
		    %10 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %11 : __torch__.models.common.BottleneckCSP = prim::GetAttr[name="20"](%10)
		    %12 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %13 : __torch__.models.common.Concat = prim::GetAttr[name="19"](%12)
		    %14 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %15 : __torch__.models.common.Conv = prim::GetAttr[name="18"](%14)
		    %16 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %17 : __torch__.models.common.BottleneckCSP = prim::GetAttr[name="17"](%16)
		    %18 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %19 : __torch__.models.common.Concat = prim::GetAttr[name="16"](%18)
		    %20 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %21 : __torch__.torch.nn.modules.upsampling.Upsample = prim::GetAttr[name="15"](%20)
		    %22 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %23 : __torch__.models.common.Conv = prim::GetAttr[name="14"](%22)
		    %24 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %25 : __torch__.models.common.BottleneckCSP = prim::GetAttr[name="13"](%24)
		    %26 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %27 : __torch__.models.common.Concat = prim::GetAttr[name="12"](%26)
		    %28 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %29 : __torch__.torch.nn.modules.upsampling.Upsample = prim::GetAttr[name="11"](%28)
		    %30 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %31 : __torch__.models.common.Conv = prim::GetAttr[name="10"](%30)
		    %32 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %33 : __torch__.models.common.BottleneckCSP = prim::GetAttr[name="9"](%32)
		    %34 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %35 : __torch__.models.common.SPP = prim::GetAttr[name="8"](%34)
		    %36 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %37 : __torch__.models.common.Conv = prim::GetAttr[name="7"](%36)
		    %38 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %39 : __torch__.models.common.BottleneckCSP = prim::GetAttr[name="6"](%38)
		    %40 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %41 : __torch__.models.common.Conv = prim::GetAttr[name="5"](%40)
		    %42 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %43 : __torch__.models.common.BottleneckCSP = prim::GetAttr[name="4"](%42)
		    %44 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %45 : __torch__.models.common.Conv = prim::GetAttr[name="3"](%44)
		    %46 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %47 : __torch__.models.common.BottleneckCSP = prim::GetAttr[name="2"](%46)
		    %48 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %49 : __torch__.models.common.Conv = prim::GetAttr[name="1"](%48)
		    %50 : __torch__.torch.nn.modules.container.Sequential = prim::GetAttr[name="model"](%self.1)
		    %51 : __torch__.models.common.Focus = prim::GetAttr[name="0"](%50)
		    %83 : bool = prim::Constant[value=1](), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.conv # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\modules\conv.py:416:0
		    %84 : bool = prim::Constant[value=0](), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.conv # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\modules\conv.py:416:0
		    %85 : None = prim::Constant(), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.conv
		    %86 : float = prim::Constant[value=0.001](), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.bn # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\functional.py:2016:0
		    %87 : float = prim::Constant[value=0.029999999999999999](), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.bn # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\functional.py:2016:0
		    %88 : float = prim::Constant[value=0.10000000000000001](), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.act # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\functional.py:1293:0
		    %89 : int = prim::Constant[value=1](), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %90 : int = prim::Constant[value=3](), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %91 : int = prim::Constant[value=9223372036854775807](), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %92 : int = prim::Constant[value=0](), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %93 : int = prim::Constant[value=2](), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %94 : __torch__.models.common.Conv = prim::GetAttr[name="conv"](%51)
		    %95 : Tensor = aten::slice(%x, %93, %92, %91, %93), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %96 : Tensor = aten::slice(%95, %90, %92, %91, %93), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %97 : Tensor = aten::slice(%x, %93, %89, %91, %93), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %98 : Tensor = aten::slice(%97, %90, %92, %91, %93), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %99 : Tensor = aten::slice(%x, %93, %92, %91, %93), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %100 : Tensor = aten::slice(%99, %90, %89, %91, %93), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %101 : Tensor = aten::slice(%x, %93, %89, %91, %93), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %102 : Tensor = aten::slice(%101, %90, %89, %91, %93), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %103 : Tensor[] = prim::ListConstruct(%96, %98, %100, %102), scope: __module.model.0
		    %input.1 : Tensor = aten::cat(%103, %89), scope: __module.model.0 # D:\zhaoyinfa\code\zdws\models\common.py:88:0
		    %105 : __torch__.torch.nn.modules.batchnorm.BatchNorm2d = prim::GetAttr[name="bn"](%94)
		    %106 : __torch__.torch.nn.modules.conv.Conv2d = prim::GetAttr[name="conv"](%94)
		    %107 : Tensor = prim::GetAttr[name="weight"](%106)
		    %108 : int[] = prim::ListConstruct(%89, %89), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.conv
		    %109 : int[] = prim::ListConstruct(%89, %89), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.conv
		    %110 : int[] = prim::ListConstruct(%89, %89), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.conv
		    %111 : int[] = prim::ListConstruct(%92, %92), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.conv
		    %input.2 : Tensor = aten::_convolution(%input.1, %107, %85, %108, %109, %110, %84, %111, %89, %83, %84, %83), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.conv # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\modules\conv.py:416:0
		    %113 : Tensor = prim::GetAttr[name="running_var"](%105)
		    %114 : Tensor = prim::GetAttr[name="running_mean"](%105)
		    %115 : Tensor = prim::GetAttr[name="bias"](%105)
		    %116 : Tensor = prim::GetAttr[name="weight"](%105)
		    %input.3 : Tensor = aten::batch_norm(%input.2, %116, %115, %114, %113, %84, %87, %86, %83), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.bn # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\functional.py:2016:0
		    %input.4 : Tensor = aten::leaky_relu_(%input.3, %88), scope: __module.model.0/__module.model.0.conv/__module.model.0.conv.act # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\functional.py:1293:0
		    %119 : float = prim::Constant[value=0.10000000000000001](), scope: __module.model.1/__module.model.1.act # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\functional.py:1293:0
		    %120 : float = prim::Constant[value=0.029999999999999999](), scope: __module.model.1/__module.model.1.bn # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\functional.py:2016:0
		    %121 : float = prim::Constant[value=0.001](), scope: __module.model.1/__module.model.1.bn # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\functional.py:2016:0
		    %122 : None = prim::Constant(), scope: __module.model.1/__module.model.1.conv
		    %123 : int = prim::Constant[value=2](), scope: __module.model.1/__module.model.1.conv # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\modules\conv.py:416:0
		    %124 : int = prim::Constant[value=1](), scope: __module.model.1/__module.model.1.conv # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\modules\conv.py:416:0
		    %125 : bool = prim::Constant[value=0](), scope: __module.model.1/__module.model.1.conv # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\modules\conv.py:416:0
		    %126 : int = prim::Constant[value=0](), scope: __module.model.1/__module.model.1.conv # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\modules\conv.py:416:0
		    %127 : bool = prim::Constant[value=1](), scope: __module.model.1/__module.model.1.conv # D:\ProgramData\Anaconda3\lib\site-packages\torch\nn\modules\conv.py:416:0
		    %128 : __torch__.torch.nn.modules.batchnorm.BatchNorm2d = prim::GetAttr[name="bn"](%49)
		    %129 : __torch__.torch.nn.modules.conv.Conv2d = prim::GetAttr[name="conv"](%49)
		    %130 : Tensor = prim::GetAttr[name="weight

**Comment by glenn-jocher** [comment]
👋 Hello, thank you for your interest in our work! This issue seems to lack the minimum requirements for a proper response, or is insufficiently detailed for us to help you. Please note that most technical problems are due to:

- **Your modified or out-of-date code**. If your issue is not reproducible in a new `git clone` version of this repo we can not debug it. Before going further run this code and verify your issue persists:
```bash
$ git clone https://github.com/ultralytics/yolov5 yolov5_new  # clone latest
$ cd yolov5_new
$ python detect.py  # verify detection

# CODE TO REPRODUCE YOUR ISSUE HERE
```

- **Your custom data.** If your issue is not reproducible in one of our 3 common datasets ([COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [COCO128](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), or [VOC](https://github.com/ultralytics/yolov5/blob/master/data/voc.yaml)) we can not debug it. Visit our [Custom Training Tutorial](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data) for guidelines on training your custom data. Examine `train_batch0.jpg` and `test_batch0.jpg` for a sanity check of your labels and images.

- **Your environment.** If your issue is not reproducible in one of the verified environments below we can not debug it. If you are running YOLOv5 locally, verify your environment meets all of the [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies specified below. If in doubt, download Python 3.8.0 from https://www.python.org/, create a new [venv](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/), and install requirements.

If none of these apply to you, we suggest you close this issue and raise a new one using the 🐛 **Bug Report template**, providing screenshots and a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) of your issue. Thank you!


## Requirements

Python 3.8 or later with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies installed, including `torch>=1.7`. To install run:
```bash
$ pip install -r requirements.txt
```


## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>


## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. These tests evaluate proper operation of basic YOLOv5 functionality, including training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([test.py](https://github.com/ultralytics/yolov5/blob/master/test.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/models/export.py)) on MacOS, Windows, and Ubuntu.

**Comment by SkalskiP** [comment]
@glenn-jocher any progress on that issue? I have the same issue.

Interesting update: If I run `ts = torch.jit.trace(model, img)` once again the last error does not occur and the tracing is successful. But this traced model doesn't work.

**Comment by glenn-jocher** [comment]
@SkalskiP @wyuzyf I looked into this. The detect() layer checks the input size to adapt the grid if required, and tracing does not seem to like this shape check (even if the shape is fine and no grid recomputation is required). The following will warn:
https://github.com/ultralytics/yolov5/blob/0cae7576a9241110157cd154fc2237e703c2719e/train.py#L335

I found a solution below. This is a YOLOv5s model displayed in TensorBoard. You can see the Detect() layer merging the 3 layers into a single output.
```python
tb_writer.add_graph(torch.jit.trace(model, imgs, strict=False), imgs)
```
<img width="893" alt="Screenshot 2021-04-11 at 01 10 09" src="https://user-images.githubusercontent.com/26833433/114286928-349bd600-9a63-11eb-941f-7139ee6cd602.png">

**Comment by glenn-jocher** [comment]
@SkalskiP @wyuzyf this bug should be resolved now in PR #2758. Please let us know if this resolves the problem on your side, and let us know if you run into any other issues!
